File: web/ViewModels/TopicViewModel.cs
using web.Models;
using web.Services;

namespace web.ViewModels
{
    public enum TopicState { Idle, Loading, Success, Error}
    public class TopicViewModel
    {
        private readonly ApiService _apiService;
        private readonly LocalStorageService _localStorage;

        public TopicState State { get; private set; } = TopicState.Idle;
        public List<TopicModel> AvailableTopics { get; private set; } = new();
        public List<TopicModel> SelectedTopics { get; private set; } = new();
        public string ErrorMessage { get; private set; } = string.Empty;
        public bool HasEnoughTopics => SelectedTopics.Count >= 3;

        public event Action? onChange;

        public TopicViewModel(ApiService apiService, LocalStorageService localStorage)
        {
            _apiService = apiService;
            _localStorage = localStorage;
        }

        public async Task FetchAvailableTopicsAsync()
        {
            SetState(TopicState.Loading);
            
            try
            {
                AvailableTopics = await _apiService.GetTopicsAsync();
                SetState(TopicState.Success);
            } catch (Exception ex)
            {
                ErrorMessage = ex.Message;
                SetState(TopicState.Error);
            }
        }
        public async Task LoadSelectedTopicsAsync()
        {
            var savedKeys = await _localStorage.GetSelectedTopicsAsync();
            SelectedTopics = AvailableTopics.Where(t => savedKeys.Contains(t.Key)).ToList();
            NotifyStateChanged();
        }
        public void ToggleTopic(TopicModel topic)
        {
            if (SelectedTopics.Any(t => t.Key == topic.Key))
            {
                SelectedTopics = SelectedTopics.Where(t => t.Key != topic.Key).ToList();
            }else
            {
                SelectedTopics = new List<TopicModel>(SelectedTopics) { topic };
            }
            NotifyStateChanged();
        }
        public bool isSelected(TopicModel topic)
        {
            return SelectedTopics.Any(t => t.Key == topic.Key);
        }

        public async Task<bool> SaveSelectedTopicsAsync()
        {
            if (!HasEnoughTopics)
            {
                ErrorMessage = "Please select at least 3 topics";
                SetState(TopicState.Error);
                return false;
            }
            SetState(TopicState.Loading);

            try
            {
                var topicKeys = SelectedTopics.Select(t => t.Key).ToList();
                await _localStorage.SaveSelectedTopicsAsync(topicKeys);
                await _apiService.UpdateTopicsAsync(topicKeys);
                SetState(TopicState.Success);
                return true;

            } catch (Exception ex)
            {
                ErrorMessage = ex.Message;
                SetState(TopicState.Error);
                return false;
            }
        }
        public async Task<bool> InitUserAsync(string deviceId)
        {
            if (!HasEnoughTopics)
            {
                ErrorMessage = "Please select at least 3 topics";
                SetState(TopicState.Error);
                return false;
            }
            SetState(TopicState.Loading);
            
            try
            {
                var topicKeys = SelectedTopics.Select(t => t.Key).ToList();
                await _apiService.InitUserAsync(deviceId, topicKeys);
                SetState(TopicState.Success);
                return true;
            } catch(Exception ex)
            {
                ErrorMessage = ex.Message;
                SetState(TopicState.Error);
                return false;
            }
        }
        private void SetState(TopicState state)
        {
            state = State;
            NotifyStateChanged();
        }
        private void NotifyStateChanged() => onChange?.Invoke();
    }
}

File: web/js/viewmodels/TopicViewModel.js
import { getTopics, initUser, updateTopics } from '../services/apiService.js';

export class TopicViewModel {
    #availableTopics = [];
    #selectedKeys = new Set();
    #state = 'idle';
    #errorMessage = '';
    #listeners = new Set();

    get availableTopics() { return this.#availableTopics; }
    get selectedKeys() { return [...this.#selectedKeys]; }
    get selectedCount() { return this.#selectedKeys.size;}
    get state() { return this.#state; }
    get errorMessage() { return this.#errorMessage; }
    get hasEnoughTopics() { return this.#selectedKeys.size >= 3; }

    subscribe(fn) {
        this.#listeners.add(fn);
        return () => this.#listeners.delete(fn);
    }

    #notify() {
        this.#listeners.forEach((fn) => fn(this));
    }

    #setState(state) {
        this.#state = state;
        this.#notify();
    }

    async fetchTopics() {
        this.#setState('loading');
        try {
            this.#availableTopics = await getTopics();
            this.#setState('success');
        } catch(err) {
            this.#errorMessage = err.message;
            this.#setState('error');
        }
    }

    toggleTopic(key) {
        if (this.#selectedKeys.has(key)) {
            this.#selectedKeys.delete(key);
        } else {
            this.#selectedKeys.add(key);
        }
        this.#notify();
    }

    isSelected(key) {
        return this.#selectedKeys.has(key);
    }

    loadSavedTopics(savedKeys = []) {
        this.#selectedKeys = new Set(savedKeys)
        this.#notify();
    }

    async initUser(deviceId) {
        if(!this.hasEnoughTopics) {
            this.#errorMessage = 'Please select at least 3 topics';
            this.#setState('error');
            return false;
        }

        this.#setState('loading');
        try {
            await initUser(deviceId, [...this.#selectedKeys]);
            this.#setState('success');
            return true;
        } catch(err) {
            this.#errorMessage = err.message;
            this.#setState('error');
            return false;
        }
    }

    async saveTopics() {
        if (!this.hasEnoughTopics) {
            this.#errorMessage = 'Please select at least 3 topics';
            this.#setState('error');
            return false;
        }

        this.#setState('loading');
        try {
            await updateTopics([...this.#selectedKeys]);
            this.#setState('success');
            return true;
        } catch (err) {
            this.#errorMessage = err.message;
            this.#setState('error');
            return false;
        }
    }
}